src/httpauth/digest.c

```diff
@@ -591,3 +613,405 @@ int httpauth_digest_chall_request_full(struct httpauth_digest_chall_req **preq,
 
 	return err;
 }
+
+
+static void httpauth_digest_response_destructor(void *arg)
+{
+	struct httpauth_digest_enc_resp *resp = arg;
+
+	mem_deref(resp->realm);
+	mem_deref(resp->nonce);
+	mem_deref(resp->opaque);
+	mem_deref(resp->algorithm);
+	mem_deref(resp->qop);
+	mem_deref(resp->response);
+	mem_deref(resp->username);
+	mem_deref(resp->username_star);
+	mem_deref(resp->uri);
+	mem_deref(resp->cnonce);
+	mem_deref(resp->nc);
+	mem_deref(resp->charset);
+}
+
+
+static int digest_response(struct httpauth_digest_enc_resp *resp,
+	const struct httpauth_digest_chall *chall,
+	const struct pl *method, const char *user,
+	const char *passwd, const char *entitybody)
+{
+	uint8_t *hash1 = NULL;
+	uint8_t *hash2 = NULL;
+	struct mbuf *mb = NULL;
+	size_t hashstringl = (resp->hash_length * 2) + 1;
+	int err = 0, n = 0;
+
+	mb = mbuf_alloc(str_len(user) + str_len(passwd) + chall->realm.l + 2);
+	if (!mb)
+		return ENOMEM;
+
+	hash1 = mem_zalloc(resp->hash_length, NULL);
+	hash2 = mem_zalloc(resp->hash_length, NULL);
+	if (!resp->response)
+		resp->response = mem_zalloc(hashstringl, NULL);
+
+	if (!resp->response || !hash1 || !hash2) {
+		err = ENOMEM;
+		goto out;
+	}
+
+	/* HASH A2 */
+	if (str_isset(resp->qop) && str_str(resp->qop, "auth-int")) {
+		if (!entitybody || str_casecmp(entitybody, "") == 0) {
+			resp->hash_function((uint8_t *)"", str_len(""), hash1);
```

Comment:
perhaps check if `hash_function` is set, otherwise it might crash.

perhaps also use the "handler" paradigm with a trailing `h`.

Reply:
i will change to the "handler" paradigm. 
The check does not hurt so i will add it. Since `digest_response` is an static function, it gets called only after an successful set of the `hash_function`. That is the reason i didn't include the check.